lints/lint_subject_contains_reserved_arpa_ip.go

```diff
@@ -0,0 +1,234 @@
+		buf.WriteString(labels[i])
+		buf.WriteString(labels[i-1])
+		buf.WriteString(labels[i-2])
+		buf.WriteString(labels[i-3])
+		if i > 4 {
+			buf.WriteString(":")
+		}
+	}
+	return net.ParseIP(buf.String())
+}
+
+// lintReversedIPAddress lints the given name as either a reversed IPv4 or IPv6
+// address under the respective ARPA zone based on the address class. An error
+// is returned if:
+//
+// 1. There aren't enough labels in the name after removing the relevant arpa
+//    suffix.
+// 2. The IP address labels don't parse as an IP address after reversing them.
+// 3. The IP address labels parse as an IP of the wrong address class for the
+//    arpa suffix the name is using.
+// 4. The IP address is within an IANA reserved range.
+func lintReversedIPAddress(name string, ipv6 bool) error {
+	numRequiredLabels := rdnsIPv4Labels
+	zoneSuffix := rdnsIPv4Suffix
+
+	if ipv6 {
+		numRequiredLabels = rdnsIPv6Labels
+		zoneSuffix = rdnsIPv6Suffix
+	}
+
+	// Strip off the zone suffix to get only the reversed IP address
+	ipName := strings.TrimSuffix(name, zoneSuffix)
+
+	// A well encoded IPv4 or IPv6 reverse DNS name will have the correct number
+	// of labels to express the address
+	ipLabels := strings.Split(ipName, ".")
+	if len(ipLabels) != numRequiredLabels {
+		return fmt.Errorf(
+			"name %q has too few leading labels (%d vs %d) to be a reverse DNS entry "+
+				"in the %q zone.",
+			name, len(ipLabels), numRequiredLabels, zoneSuffix)
+	}
+
+	// Reverse the IP labels and try to parse an IP address
+	var ip net.IP
+	if ipv6 {
+		ip = reversedLabelsToIPv6(ipLabels)
+	} else {
+		ip = reversedLabelsToIPv4(ipLabels)
+	}
+	// If the result isn't an IP at all, that's a problem.
+	if ip == nil {
+		return fmt.Errorf(
+			"the first %d labels of name %q did not parse as a reversed IP address",
+			numRequiredLabels, name)
+	}
+
+	if !ipv6 && ip.To4() == nil {
+		// If we weren't expecting IPv6 and got it, that's a problem
+		return fmt.Errorf(
+			"the first %d labels of name %q parsed as a reversed IPv6 address but is "+
+				"in the %q IPv4 reverse DNS zone.",
+			numRequiredLabels, name, rdnsIPv4Suffix)
+	} else if ipv6 && ip.To4() != nil {
+		// If we were expecting IPv6 and got an IPv4 address, that's a problem
+		return fmt.Errorf(
+			"the first %d labels of name %q parsed as a reversed IPv4 address but is "+
+				"in the %q IPv4 reverse DNS zone.",
+			numRequiredLabels, name, rdnsIPv6Suffix)
+	}
+
+	// If the IP address is in an IANA reserved space, that's a problem.
+	if util.IsIANAReserved(ip) {
+		return fmt.Errorf(
+			"the first %d labels of name %q parsed as a reversed IP address in "+
+				"an IANA reserved IP space.",
+			numRequiredLabels, name)
+	}
+
+	return nil
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "subject_contains_reserved_arpa_ip",
+		Description:   "Checks no subject domain name contains a rDNS entry in an .arpa zone specifying a reserved IP address",
+		Citation:      "BRs: 7.1.4.2.1",
```

Comment:
Note to reviewers: I took it as an assumption that the reason issuing for reverse DNS names specifying addresses in an IANA reserved network is against the BR is the chapter and verse that forbids issuing for private IP addresses, "Section 7.1.4.2.1: Subject Alternative Name Extension".

This might not be the best reference and the `EffectiveDate` might need to be tweaked if it isn't.

Reply:
That's the best I found when going through the BRs earlier this week too. :/